src/AzureAuth/Commands/Ado/CommandToken.cs

```diff
@@ -3,59 +3,127 @@
-        public string PromptHint { get; set; }
+        /// <param name="value">The PAT value.</param>
+        /// <param name="output">The output mode.</param>
+        /// <param name="scheme">The <see cref="Authorization"/> scheme to use.</param>
+        /// <returns>The formatted PAT value ready for printing.</returns>
+        public static string FormatToken(string value, OutputMode output, Authorization scheme) => output switch
+        {
+            OutputMode.Token => value,
+            OutputMode.Header => value.AsHeader(scheme),
+            OutputMode.HeaderValue => value.AsHeaderValue(scheme),
+            _ => throw new ArgumentOutOfRangeException(nameof(scheme)),
+        };
 
         /// <summary>
         /// Executes the command and returns a status code indicating the success or failure of the execution.
         /// </summary>
         /// <param name="logger">The <see cref="ILogger{T}"/> instance that is used for logging.</param>
+        /// <param name="env">An <see cref="IEnv"/> to use.</param>
         /// <param name="eventData">Lasso injected command event data.</param>
         /// <returns>An integer status code. 0 for success and non-zero for failure.</returns>
-        public int OnExecute(ILogger<CommandToken> logger, CommandExecuteEventData eventData)
+        public int OnExecute(ILogger<CommandToken> logger, IEnv env, CommandExecuteEventData eventData)
         {
-            logger.LogInformation("coming soon");
-            return 0;
+            // First attempt using a PAT.
+            var pat = PatFromEnv.Get(env);
+            if (pat.Exists)
+            {
+                logger.LogDebug($"Using PAT from env var {pat.EnvVarSource}");
+                logger.LogInformation(FormatToken(pat.Value, this.Output, Authorization.Basic));
+                return 0;
+            }
+
+            // If no PAT then use AAD AT.
+            var authResult = AzureAuth.Ado.TokenFetcher.AccessToken(
+                logger: logger,
+                mode: this.AuthModes.Combine().PreventInteractionIfNeeded(env),
+                domain: this.Domain,
+                prompt: AzureAuth.PromptHint.Prefixed(this.PromptHint),
+                timeout: TimeSpan.FromMinutes(this.Timeout));
+
+            var authflow = authResult.Success;
+            if (authflow != null)
+            {
+                logger.LogDebug($"Acquired AAD AT via {authflow.AuthFlowName} in {authflow.Duration.TotalSeconds:0.00} sec");
+                logger.LogInformation(FormatToken(authflow.TokenResult.Token, this.Output, Authorization.Bearer));
+                return 0;
+            }
+
+            logger.LogError($"Failed to find a PAT and authenticate to ADO.");
+            foreach (var attempt in authResult.Attempts)
+            {
+                logger.LogError($"{attempt.AuthFlowName} failed after {attempt.Duration.TotalSeconds:0.00} sec. Error count: {attempt.Errors.Count}");

```

Comment:
Do we need to capture this? We know that if land here all authflows have failed. The timeout for each is already hard coded in code.

Reply:
It's just just printing more information about each flow. This exact logic I plan to move into a common orchestrator. This way more of the logging can be consolidated into 1 place, and the indivudal auth flows can be more concerned with executing and returning results. The caller (this level of code) can then take more ownership over how to print those results.